draft-irtf-cfrg-opaque.md

```diff
@@ -393,152 +397,342 @@ This specification defines one particular AKE based on 3DH; see {{instantiations
+The below structure represents an example of how these values might be conveniently stored together on the server.
+
+~~~
 struct {
-  InnerEnvelope inner_env;
-  opaque auth_tag[Nm];
-} Envelope;
+    SerializedScalar oprf_key;
+    opaque client_public_key[Npk];
+    Envelope envelope;
+} credential_file;
 ~~~
 
-mode
-: The `EnvelopeMode` value. This MUST be one of `base` or `custom_identifier`.
+# Offline Registration {#offline-phase}
+
+Registration is executed between a client C and a
+server S. It is assumed S can identify C and the client can
+authenticate S during this registration phase. This is the only part
+in OPAQUE that requires an authenticated and confidential channel, either physical, out-of-band,
+PKI-based, etc. This section describes the registration flow, message encoding,
+and helper functions. Moreover, C has a key pair (client_private_key, client_public_key) for an AKE protocol
+which is suitable for use with OPAQUE; See {{online-phase}}. The private-public keys (client_private_key, client_public_key) may be randomly generated (using a cryptographically secure pseudorandom number generator) for the account or provided by the calling client.
+Clients MUST NOT use the same key pair (client_private_key, client_public_key) for two different accounts.
+
+## Setup Phase {#setup-phase}
+
+In a setup phase, C chooses its password, and S chooses its own pair of private-public
+AKE keys (server_private_key, server_public_key) for use with the AKE, along with an Nh-byte oprf_seed. S can use
+the same pair of keys with multiple clients and can opt to use multiple seeds (so long as they are
+kept consistent for each client). These steps can happen offline, i.e., before the registration phase.
+
+Once complete, the registration process proceeds as follows.
+
+## Credential Registration
 
-nonce
-: A unique 32-byte nonce used to protect this Envelope.
+~~~
+ Client (password, creds)            Server (server_private_key, server_public_key, credential_identifier, oprf_seed)
+ --------------------------------------------------------------------
+ (request, blind) = CreateRegistrationRequest(password)
 
-encrypted_creds
-: Encoding of encrypted and authenticated `SecretCredentials`.
+                               request
+                      ------------------------->
+
+            (response, oprf_key) = CreateRegistrationResponse(request, server_public_key, credential_identifier, oprf_seed)
 
-auth_tag
-: Authentication tag protecting the contents of the envelope,
-covering `InnerEnvelope` and `CleartextCredentials`.
+                               response
+                      <-------------------------
 
-The full procedure for constructing `Envelope` and `InnerEnvelope` from
-`SecretCredentials` and `CleartextCredentials` is described in {{finalize-request}}.
-Note that only `SecretCredentials` are stored in the `Envelope` (in encrypted form).
+ (record, export_key) = FinalizeRequest(password, creds, blind, response)
```

Comment:
The API for FinalizeRequest should be updated (first argument is the client_private_key now)